types/moddable/modules/ble.d.ts

```diff
@@ -0,0 +1,732 @@
+         * are returned to the `onCharacteristics` callback.
+         */
+        discoverAllCharacteristics(): void;
+
+        /**
+         * Discover a single service characteristic by UUID.
+         *
+         * @param uuid
+         * @url https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/ble/ble.md#discovercharacteristicuuid
+         */
+        discoverCharacteristic(uuid: Bytes): void;
+
+        /**
+         * Called when characteristic discovery completes.
+         * If `findCharacteristicByUUID` was called to find a single characteristic,
+         * the `characteristics` array contains the single characteristic found.
+         *
+         * @param characteristics
+         */
+        onCharacteristics(characteristics: Characteristic[]): void;
+    }
+
+    /**
+     * A single service Characteristic.
+     *
+     * @url https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/ble/ble.md#class-characteristic
+     */
+    export interface Characteristic {
+        connection: number;
+        uuid: Bytes;
+        service: Service;
+        handle: number;
+        name: string;
+        type: string;
+        descriptors: Descriptor[];
+
+        /**
+         * Discover all the characteristic's descriptors.
+         * Discovered descriptors are returned to the `onDescriptors` callback.
+         */
+        discoverAllDescriptors(): void;
+
+        /**
+         * Called when descriptor discovery completes.
+         *
+         * @param descriptors
+         */
+        onDescriptors(descriptors: Descriptor[]): void;
+
+        /**
+         * Enable characteristic value change notifications.
+         *
+         * @url https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/ble/ble.md#enablenotifications
+         */
+        disableNotifications(): void;
+    }
+
+    /**
+     * A single Characteristic descriptor.
+     *
+     * @url https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/ble/ble.md#class-descriptor
+     */
+    export interface Descriptor {
+        connection: number;
+        uuid: string;
+        characteristic: Characteristic;
+        handle: number;
+        name: string;
+        type: string;
+
+        /**
+         * Read a descriptor value on demand.
+         *
+         * @param auth
+         * @url https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/ble/ble.md#readvalueauth-1
+         */
+        readValue(auth?: number): void;
+    }
+}
+
+declare module "connection" {
+    export class Connection {
+        disconnect(): void;
+    }
+}
+
+declare module "gap" {
+
+    enum gapAddressTypes {
+        PUBLIC = 0,
+        RANDOM = 1,
+        RPA_PUBLIC = 2,
+        RPA_RANDOM = 3,
+    }
+
+    export const GAP: {
+        readonly SCAN_FAST_INTERVAL: 0x0030,		// TGAP(scan_fast_interval)		30ms to 60ms
```

Comment:
<!--
  1 failure:  The root .editorc...
  0 warning: 
  
  
  DangerID: danger-id-Danger;
  File: types/moddable/modules/ble.d.ts;
  Line: 611;
-->
- :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.